File: app/src/main/java/com/yuriy/openradio/service/MediaNotification.java
/*
 * Copyright (C) 2016 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.yuriy.openradio.service;

import android.app.PendingIntent;
import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.content.pm.ApplicationInfo;
import android.content.res.Resources;
import android.content.res.TypedArray;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.Color;
import android.os.AsyncTask;
import android.os.RemoteException;
import android.support.annotation.NonNull;
import android.support.v4.app.NotificationCompat;
import android.support.v4.app.NotificationManagerCompat;
import android.support.v4.media.MediaDescriptionCompat;
import android.support.v4.media.MediaMetadataCompat;
import android.support.v4.media.session.MediaControllerCompat;
import android.support.v4.media.session.MediaSessionCompat;
import android.support.v4.media.session.PlaybackStateCompat;
import android.text.TextUtils;
import android.util.LruCache;

import com.yuriy.openradio.R;
import com.yuriy.openradio.utils.AppLogger;
import com.yuriy.openradio.utils.BitmapHelper;
import com.yuriy.openradio.utils.FabricUtils;

import java.io.IOException;
import java.lang.ref.WeakReference;

/**
 * Keeps track of a notification and updates it automatically for a given
 * MediaSession. Maintaining a visible notification (usually) guarantees that the music service
 * won't be killed during playback.
 */
public class MediaNotification extends BroadcastReceiver {
    
    private static final String CLASS_NAME = MediaNotification.class.getSimpleName();

    private static final int NOTIFICATION_ID = 412;

    private static final String ACTION_PAUSE = "com.yuriy.openradio.pause";
    private static final String ACTION_PLAY = "com.yuriy.openradio.play";
    private static final String ACTION_PREV = "com.yuriy.openradio.prev";
    private static final String ACTION_NEXT = "com.yuriy.openradio.next";

    private static final int MAX_ALBUM_ART_CACHE_SIZE = 1024 * 1024;

    private final OpenRadioService mService;
    private MediaSessionCompat.Token mSessionToken;
    private MediaControllerCompat mController;
    private MediaControllerCompat.TransportControls mTransportControls;
    private final LruCache<String, Bitmap> mAlbumArtCache;

    private PlaybackStateCompat mPlaybackState;
    private MediaMetadataCompat mMetadata;

    private NotificationCompat.Builder mNotificationBuilder;
    private NotificationManagerCompat mNotificationManager;
    private NotificationCompat.Action mPlayPauseAction;

    private PendingIntent mPauseIntent, mPlayIntent, mPreviousIntent, mNextIntent;

    private int mNotificationColor;

    private boolean mStarted = false;

    public MediaNotification(final OpenRadioService service) {
        mService = service;
        updateSessionToken();

        // simple album art cache that holds no more than
        // MAX_ALBUM_ART_CACHE_SIZE bytes:
        mAlbumArtCache = new LruCache<String, Bitmap>(MAX_ALBUM_ART_CACHE_SIZE) {
            @Override
            protected int sizeOf(String key, Bitmap value) {
                return value.getByteCount();
            }
        };

        mNotificationColor = getNotificationColor();

        mNotificationManager = NotificationManagerCompat.from(mService);

        String pkg = mService.getPackageName();
        mPauseIntent = PendingIntent.getBroadcast(mService, 100,
                new Intent(ACTION_PAUSE).setPackage(pkg), PendingIntent.FLAG_CANCEL_CURRENT);
        mPlayIntent = PendingIntent.getBroadcast(mService, 100,
                new Intent(ACTION_PLAY).setPackage(pkg), PendingIntent.FLAG_CANCEL_CURRENT);
        mPreviousIntent = PendingIntent.getBroadcast(mService, 100,
                new Intent(ACTION_PREV).setPackage(pkg), PendingIntent.FLAG_CANCEL_CURRENT);
        mNextIntent = PendingIntent.getBroadcast(mService, 100,
                new Intent(ACTION_NEXT).setPackage(pkg), PendingIntent.FLAG_CANCEL_CURRENT);
    }

    private int getNotificationColor() {
        int notificationColor = 0;
        String packageName = mService.getPackageName();
        try {
            Context packageContext = mService.createPackageContext(packageName, 0);
            ApplicationInfo applicationInfo =
                    mService.getPackageManager().getApplicationInfo(packageName, 0);
            packageContext.setTheme(applicationInfo.theme);
            Resources.Theme theme = packageContext.getTheme();
            TypedArray ta = theme.obtainStyledAttributes(
                    new int[] {android.support.v7.appcompat.R.attr.colorPrimary});
            notificationColor = ta.getColor(0, Color.DKGRAY);
            ta.recycle();
        } catch (final Exception e) {
            FabricUtils.logException(e);
        }
        return notificationColor;
    }

    /**
     * Posts the notification and starts tracking the session to keep it
     * updated. The notification will automatically be removed if the session is
     * destroyed before {@link #stopNotification} is called.
     */
    public void startNotification() {
        if (!mStarted) {
            mController.registerCallback(mCb);
            IntentFilter filter = new IntentFilter();
            filter.addAction(ACTION_NEXT);
            filter.addAction(ACTION_PAUSE);
            filter.addAction(ACTION_PLAY);
            filter.addAction(ACTION_PREV);
            mService.registerReceiver(this, filter);

            mMetadata = mController.getMetadata();
            mPlaybackState = mController.getPlaybackState();

            mStarted = true;
            // The notification must be updated after setting started to true
            updateNotificationMetadata();
        }
    }

    /**
     * Removes the notification and stops tracking the session. If the session
     * was destroyed this has no effect.
     */
    public void stopNotification() {
        mStarted = false;
        mController.unregisterCallback(mCb);
        try {
            mNotificationManager.cancel(NOTIFICATION_ID);
            mService.unregisterReceiver(this);
        } catch (final IllegalArgumentException ex) {
            // ignore if the receiver is not registered.
        }
        mService.stopForeground(true);
    }

    @Override
    public void onReceive(final Context context, final Intent intent) {
        final String action = intent.getAction();
        AppLogger.d(CLASS_NAME + " Received intent with action " + action);
        if (TextUtils.isEmpty(action)) {
            return;
        }
        switch (action) {
            case ACTION_PAUSE:
                mTransportControls.pause();
                break;
            case ACTION_PLAY:
                mTransportControls.play();
                break;
            case ACTION_NEXT:
                mTransportControls.skipToNext();
                break;
            case ACTION_PREV:
                mTransportControls.skipToPrevious();
                break;
        }
    }

    /**
     * Update the state based on a change on the session token. Called either when
     * we are running for the first time or when the media session owner has destroyed the session
     * (see {@link android.media.session.MediaController.Callback#onSessionDestroyed()})
     */
    private void updateSessionToken() {
        final MediaSessionCompat.Token freshToken = mService.getSessionToken();
        if (mSessionToken == null || !mSessionToken.equals(freshToken)) {
            if (mController != null) {
                mController.unregisterCallback(mCb);
            }
            mSessionToken = freshToken;
            try {
                mController = new MediaControllerCompat(mService, mSessionToken);
            } catch (final RemoteException e) {
                FabricUtils.logException(e);
                return;
            }
            mTransportControls = mController.getTransportControls();
            if (mStarted) {
                mController.registerCallback(mCb);
            }
        }
    }

    private final MediaControllerCompat.Callback mCb = new MediaControllerCompat.Callback() {

        @Override
        public void onPlaybackStateChanged(@NonNull PlaybackStateCompat state) {
            mPlaybackState = state;
            AppLogger.d(CLASS_NAME + " Received new playback state:" + state);
            updateNotificationPlaybackState();
        }

        @Override
        public void onMetadataChanged(MediaMetadataCompat metadata) {
            mMetadata = metadata;
            AppLogger.d(CLASS_NAME + " Received new metadata:" + metadata);
            updateNotificationMetadata();
        }

        @Override
        public void onSessionDestroyed() {
            super.onSessionDestroyed();
            AppLogger.d(CLASS_NAME + " Session was destroyed, resetting to the new session token");
            updateSessionToken();
        }
    };

    private void updateNotificationMetadata() {
        AppLogger.d(CLASS_NAME + " Update Notification Metadata : " + mMetadata);
        if (mMetadata == null || mPlaybackState == null) {
            return;
        }

        updatePlayPauseAction();

        mNotificationBuilder = new NotificationCompat.Builder(mService);
        int playPauseActionIndex = 0;

        // If skip to previous action is enabled
        if ((mPlaybackState.getActions() & PlaybackStateCompat.ACTION_SKIP_TO_PREVIOUS) != 0) {
            mNotificationBuilder
                    .addAction(R.drawable.ic_skip_previous_white_24dp,
                            mService.getString(R.string.label_previous), mPreviousIntent);
            playPauseActionIndex = 1;
        }

        mNotificationBuilder.addAction(mPlayPauseAction);

        // If skip to next action is enabled
        if ((mPlaybackState.getActions() & PlaybackStateCompat.ACTION_SKIP_TO_NEXT) != 0) {
            mNotificationBuilder.addAction(R.drawable.ic_skip_next_white_24dp,
                    mService.getString(R.string.label_next), mNextIntent);
        }

        final MediaDescriptionCompat description = mMetadata.getDescription();

        String fetchArtUrl = null;
        Bitmap art = description.getIconBitmap();
        if (art == null && description.getIconUri() != null) {
            // This sample assumes the iconUri will be a valid URL formatted String, but
            // it can actually be any valid Android Uri formatted String.
            // async fetch the album art icon
            final String artUrl = description.getIconUri().toString();
            art = mAlbumArtCache.get(artUrl);
            if (art == null) {
                fetchArtUrl = artUrl;
                // use a placeholder art while the remote art is being downloaded
                art = BitmapFactory.decodeResource(mService.getResources(), R.drawable.ic_radio_station);
            }
        }

        mNotificationBuilder
                .setStyle(new android.support.v4.media.app.NotificationCompat.MediaStyle()
                        .setShowActionsInCompactView(playPauseActionIndex)  // only show play/pause in compact view
                        .setMediaSession(mSessionToken))
                .setColor(mNotificationColor)
                .setSmallIcon(R.drawable.ic_notification)
                .setVisibility(NotificationCompat.VISIBILITY_PUBLIC)
                .setUsesChronometer(true)
                .setContentTitle(description.getTitle())
                .setContentText(description.getSubtitle())
                .setLargeIcon(art);

        updateNotificationPlaybackState();

        mService.startForeground(NOTIFICATION_ID, mNotificationBuilder.build());
        if (fetchArtUrl != null && !BitmapHelper.isUrlLocalResource(fetchArtUrl)) {
            fetchBitmapFromURLAsync(fetchArtUrl);
        }
    }

    private void updatePlayPauseAction() {
        AppLogger.d(CLASS_NAME + " updatePlayPauseAction");
        String label;
        int icon;
        PendingIntent intent;
        if (mPlaybackState.getState() == PlaybackStateCompat.STATE_PLAYING) {
            label = mService.getString(R.string.label_pause);
            icon = R.drawable.ic_pause_white_24dp;
            intent = mPauseIntent;
        } else {
            label = mService.getString(R.string.label_play);
            icon = R.drawable.ic_play_arrow_white_24dp;
            intent = mPlayIntent;
        }
        if (mPlayPauseAction == null) {
            mPlayPauseAction = new NotificationCompat.Action(icon, label, intent);
        } else {
            mPlayPauseAction.icon = icon;
            mPlayPauseAction.title = label;
            mPlayPauseAction.actionIntent = intent;
        }
    }

    private void updateNotificationPlaybackState() {
        AppLogger.d(CLASS_NAME + " updateNotificationPlaybackState. mPlaybackState=" + mPlaybackState);
        if (mPlaybackState == null || !mStarted) {
            AppLogger.d(CLASS_NAME + " updateNotificationPlaybackState. cancelling notification!");
            mService.stopForeground(true);
            return;
        }
        if (mNotificationBuilder == null) {
            AppLogger.d(CLASS_NAME + " updateNotificationPlaybackState. there is no notificationBuilder. Ignoring request to update state!");
            return;
        }
        if (mPlaybackState.getPosition() >= 0) {
            AppLogger.d(CLASS_NAME + " updateNotificationPlaybackState. updating playback position to " +
                    (System.currentTimeMillis() - mPlaybackState.getPosition()) / 1000 + " seconds");
            mNotificationBuilder
                    .setWhen(System.currentTimeMillis() - mPlaybackState.getPosition())
                    .setShowWhen(true)
                    .setUsesChronometer(true);
            mNotificationBuilder.setShowWhen(true);
        } else {
            AppLogger.d(CLASS_NAME + " updateNotificationPlaybackState. hiding playback position");
            mNotificationBuilder
                    .setWhen(0)
                    .setShowWhen(false)
                    .setUsesChronometer(false);
        }

        updatePlayPauseAction();

        // Make sure that the notification can be dismissed by the user when we are not playing:
        mNotificationBuilder.setOngoing(mPlaybackState.getState() == PlaybackStateCompat.STATE_PLAYING);

        mNotificationManager.notify(NOTIFICATION_ID, mNotificationBuilder.build());
    }

    private void fetchBitmapFromURLAsync(final String source) {
        AppLogger.d(CLASS_NAME + " getBitmapFromURLAsync: starting async task to fetch " + source);
        final AsyncTask<Void, Void, Bitmap> task = new FetchBitmapAsyncTask(this, source);
        task.execute();
    }

    /**
     * async task to fetch Bitmap from provided URL.
     */
    private static final class FetchBitmapAsyncTask extends AsyncTask<Void, Void, Bitmap> {

        /**
         * Reference to the enclosing class.
         */
        private final WeakReference<MediaNotification> mReference;

        /**
         * Resource URL.
         */
        private final String mSource;

        /**
         * Main constructor.
         *
         * @param mediaNotification Reference to the enclosing class.
         * @param source            URL of the resource.
         */
        private FetchBitmapAsyncTask(final MediaNotification mediaNotification,
                                     final String source) {
            super();
            mReference = new WeakReference<>(mediaNotification);
            mSource = source;
        }

        @Override
        protected Bitmap doInBackground(final Void... params) {
            final MediaNotification reference = mReference.get();
            if (reference == null) {
                return  null;
            }

            Bitmap bitmap = null;
            try {
                bitmap = BitmapHelper.fetchAndRescaleBitmap(
                        mSource,
                        BitmapHelper.MEDIA_ART_BIG_WIDTH,
                        BitmapHelper.MEDIA_ART_BIG_HEIGHT
                );
            } catch (final IOException e) {
                FabricUtils.logException(e);
            }
            if (mSource != null && bitmap != null) {
                reference.mAlbumArtCache.put(mSource, bitmap);
            } else {
                AppLogger.e(CLASS_NAME + " Can not put data into AlbumArtCache, " +
                        "key == null || value == null");
            }
            return bitmap;
        }

        @Override
        protected void onPostExecute(final Bitmap bitmap) {
            final MediaNotification reference = mReference.get();
            if (reference == null) {
                return;
            }
            if (bitmap != null && reference.mMetadata != null
                    && reference.mNotificationBuilder != null
                    && reference.mMetadata.getDescription().getIconUri() != null
                    && !TextUtils.equals(mSource, reference.mMetadata.getDescription().getIconUri().toString())) {
                // If the media is still the same, update the notification:
                AppLogger.d(CLASS_NAME + " GetBitmapFromURLAsync: set bitmap to " + mSource);
                reference.mNotificationBuilder.setLargeIcon(bitmap);
                reference.mNotificationManager.notify(
                        NOTIFICATION_ID, reference.mNotificationBuilder.build()
                );
            }
        }
    }
}

File: common/src/main/java/com/yuriy/openradio/shared/notification/MediaNotification.kt
/*
 * Copyright (C) 2016-2021 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.yuriy.openradio.shared.notification

import android.app.PendingIntent
import android.content.BroadcastReceiver
import android.content.ComponentName
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.graphics.BitmapFactory
import android.graphics.Color
import android.os.RemoteException
import android.support.v4.media.MediaMetadataCompat
import android.support.v4.media.session.MediaControllerCompat
import android.support.v4.media.session.MediaSessionCompat
import android.support.v4.media.session.PlaybackStateCompat
import androidx.core.app.NotificationCompat
import androidx.core.app.NotificationManagerCompat
import com.yuriy.openradio.R
import com.yuriy.openradio.shared.service.OpenRadioService
import com.yuriy.openradio.shared.utils.AppLogger
import com.yuriy.openradio.shared.utils.AppUtils
import com.yuriy.openradio.shared.utils.MediaItemHelper
import com.yuriy.openradio.shared.vo.RadioStation
import java.util.*
import java.util.concurrent.atomic.AtomicBoolean

/**
 * Keeps track of a notification and updates it automatically for a given
 * MediaSession. Maintaining a visible notification (usually) guarantees that the music service
 * won't be killed during playback.
 *
 * @param mContext Application context.
 * @param mService Open Radio Service.
 */
class MediaNotification(private val mContext: Context, private val mService: OpenRadioService) : BroadcastReceiver() {

    private lateinit var mSessionToken: MediaSessionCompat.Token
    private lateinit var mController: MediaControllerCompat
    private var mTransportControls: MediaControllerCompat.TransportControls? = null
    private var mMetadata: MediaMetadataCompat? = null
    private val mCb: MediaControllerCompatCallback
    private val mNotificationManager: NotificationManagerCompat
    private val mPauseIntent: PendingIntent
    private val mPlayIntent: PendingIntent
    private val mPreviousIntent: PendingIntent
    private val mNextIntent: PendingIntent
    private val mCloseAppIntent: PendingIntent
    private val mNotificationColor: Int
    private val mStarted = AtomicBoolean(false)
    private val mNotificationChannelFactory: NotificationChannelFactory
    private val mUseNavigationActionsInCompactView = false
    private val notificationColor: Int
        get() {
            var notificationColor = 0
            val packageName = mContext.packageName
            try {
                val packageContext = mContext.createPackageContext(packageName, 0)
                val applicationInfo = mContext.packageManager.getApplicationInfo(packageName, 0)
                packageContext.setTheme(applicationInfo.theme)
                val theme = packageContext.theme
                if (AppUtils.hasVersionLollipop()) {
                    val ta = theme.obtainStyledAttributes(intArrayOf(R.attr.colorPrimary))
                    notificationColor = ta.getColor(0, Color.DKGRAY)
                    ta.recycle()
                } else {
                    notificationColor = -0xbbbbbc
                }
            } catch (e: Exception) {
                AppLogger.e("Get notification color", e)
            }
            return notificationColor
        }

    init {
        mCb = MediaControllerCompatCallback()
        mNotificationChannelFactory = NotificationChannelFactory(mContext)
        updateSessionToken()

        mNotificationColor = notificationColor
        mNotificationManager = NotificationManagerCompat.from(mContext)
        val pkg = mContext.packageName
        mPauseIntent = PendingIntent.getBroadcast(
            mContext, 100,
            Intent(ACTION_PAUSE).setPackage(pkg), PendingIntent.FLAG_IMMUTABLE
        )
        mPlayIntent = PendingIntent.getBroadcast(
            mContext, 100,
            Intent(ACTION_PLAY).setPackage(pkg), PendingIntent.FLAG_IMMUTABLE
        )
        mPreviousIntent = PendingIntent.getBroadcast(
            mContext, 100,
            Intent(ACTION_PREV).setPackage(pkg), PendingIntent.FLAG_IMMUTABLE
        )
        mNextIntent = PendingIntent.getBroadcast(
            mContext, 100,
            Intent(ACTION_NEXT).setPackage(pkg), PendingIntent.FLAG_IMMUTABLE
        )
        mCloseAppIntent = PendingIntent.getBroadcast(
            mContext, 100,
            Intent(ACTION_CLOSE_APP).setPackage(pkg), PendingIntent.FLAG_IMMUTABLE
        )
    }

    /**
     * Posts the notification and starts tracking the session to keep it
     * updated. The notification will automatically be removed if the session is
     * destroyed before [.stopNotification] is called.
     */
    fun startNotification(context: Context, radioStation: RadioStation) {
        if (mStarted.get()) {
            return
        }
        mController.registerCallback(mCb)
        val filter = IntentFilter()
        filter.addAction(ACTION_NEXT)
        filter.addAction(ACTION_PAUSE)
        filter.addAction(ACTION_PLAY)
        filter.addAction(ACTION_PREV)
        filter.addAction(ACTION_CLOSE_APP)
        context.registerReceiver(this, filter)
        val metadata = mController.metadata
        mMetadata = metadata ?: MediaItemHelper.metadataFromRadioStation(context, radioStation)
        mStarted.set(true)
        // The notification must be updated after setting started to true
        handleNotification(mController.playbackState)
    }

    /**
     * Removes the notification and stops tracking the session. If the session
     * was destroyed this has no effect.
     */
    fun stopNotification(context: Context) {
        if (!mStarted.get()) {
            return
        }
        mController.unregisterCallback(mCb)
        mNotificationManager.cancelAll()
        try {
            context.unregisterReceiver(this)
        } catch (ex: IllegalArgumentException) {
            AppLogger.e("$CLASS_NAME error while unregister", ex)
        }
        mStarted.set(false)
    }

    override fun onReceive(context: Context, intent: Intent) {
        AppLogger.d("$CLASS_NAME received intent $intent")
        when (intent.action) {
            ACTION_PAUSE -> mTransportControls!!.pause()
            ACTION_PLAY -> mTransportControls!!.play()
            ACTION_NEXT -> mTransportControls!!.skipToNext()
            ACTION_PREV -> mTransportControls!!.skipToPrevious()
            ACTION_CLOSE_APP -> {
                AppLogger.i("$CLASS_NAME close App from Notification")
                mService.closeService()
            }
        }
    }

    /**
     * Update the state based on a change on the session token. Called either when
     * we are running for the first time or when the media session owner has destroyed the session
     * (see [android.media.session.MediaController.Callback.onSessionDestroyed])
     */
    private fun updateSessionToken() {
        val freshToken = mService.sessionToken
        AppLogger.d("$CLASS_NAME fresh token:$freshToken")
        if (!this::mSessionToken.isInitialized || mSessionToken != freshToken) {
            if (this::mController.isInitialized) {
                mController.unregisterCallback(mCb)
            }
            if (freshToken != null) {
                mSessionToken = freshToken
            }
            mController = try {
                MediaControllerCompat(mContext, mSessionToken)
            } catch (e: RemoteException) {
                AppLogger.e("UpdateSessionToken", e)
                return
            }
            mTransportControls = mController.transportControls
            if (mStarted.get()) {
                mController.registerCallback(mCb)
            }
        }
    }

    private inner class MediaControllerCompatCallback : MediaControllerCompat.Callback() {

        private var mPlaybackState: PlaybackStateCompat? = null

        override fun onPlaybackStateChanged(state: PlaybackStateCompat) {
            AppLogger.d("$CLASS_NAME Received new playback state:$state")
            val doNotify = if (mPlaybackState == null) {
                true
            } else {
                doHandleState(mPlaybackState!!.state, state.state)
            }
            if (doNotify) {
                handleNotification(state)
            }
            mPlaybackState = state
        }

        override fun onMetadataChanged(metadata: MediaMetadataCompat) {
            AppLogger.d("$CLASS_NAME Received new metadata:$metadata")
            mMetadata = metadata
            handleNotification(mPlaybackState)
        }

        override fun onSessionDestroyed() {
            AppLogger.d("$CLASS_NAME Session was destroyed, resetting to the new session token")
            updateSessionToken()
        }

        private fun doHandleState(curState: Int, newState: Int): Boolean {
            return newState != curState
        }
    }

    fun handleNotification(playbackState: PlaybackStateCompat?) {
        if (mMetadata == null) {
            return
        }
        if (playbackState == null) {
            return
        }
        val builder = NotificationCompat.Builder(
            mContext, MediaNotificationData.CHANNEL_ID
        )

        // Create/Retrieve Notification Channel for O and beyond devices (26+).
        mNotificationChannelFactory.createChannel(MediaNotificationData(mContext, mMetadata!!))
        // If skip to previous action is enabled
        var enablePrevious = false
        if (playbackState.actions and PlaybackStateCompat.ACTION_SKIP_TO_PREVIOUS != 0L) {
            builder.addAction(
                R.drawable.ic_skip_prev,
                mContext.getString(R.string.label_previous),
                mPreviousIntent
            )
            enablePrevious = true
        }

        val description = mMetadata!!.description
        var art = description.iconBitmap
        if (art == null && description.iconUri != null) {
            //art = description.iconUri
        } else if (art == null) {
            // use a placeholder art while the remote art is being downloaded
            art = BitmapFactory.decodeResource(mContext.resources, R.drawable.ic_radio_station)
        } else {
            AppLogger.d("Art bitmap:$art")
        }

        builder.addAction(getPlayPauseAction(playbackState))

        // If skip to next action is enabled
        var enableNext = false
        if (playbackState.actions and PlaybackStateCompat.ACTION_SKIP_TO_NEXT != 0L) {
            builder.addAction(
                R.drawable.ic_skip_next,
                mContext.getString(R.string.label_next),
                mNextIntent
            )
            enableNext = true
        }
        val smallIcon =
            if (AppUtils.hasVersionLollipop()) R.drawable.ic_notification else R.drawable.ic_notification_drawable
        // Build the style.
        val actionsToShowInCompact = getActionIndicesForCompactView(
            getActions(enableNext, enablePrevious, playbackState), playbackState
        )
        val mediaStyle = androidx.media.app.NotificationCompat.MediaStyle()
            .setShowActionsInCompactView(*actionsToShowInCompact)
            .setMediaSession(mSessionToken)
        builder
            .setContentIntent(makePendingIntent())
            .setStyle(mediaStyle)
            .setColor(mNotificationColor)
            .setLargeIcon(art)
            .setSmallIcon(smallIcon)
            .setOngoing(true)
            .setDefaults(0)
            .setSound(null)
            .setVisibility(NotificationCompat.VISIBILITY_PUBLIC)
            .setContentTitle(description.title)
            .setContentText(description.subtitle)
        AppLogger.d(
            CLASS_NAME + " Update Notification for ${description.mediaId} " +
                "state:" + playbackState +
                " title:" + description.title +
                " subtitle:" + description.subtitle
        )

        builder.addAction(getCloseAppAction())

        mService.startForeground(NOTIFICATION_ID, builder.build())
    }

    fun notifyService(message: String) {
        val art = BitmapFactory.decodeResource(
            mContext.resources, R.drawable.ic_radio_station
        )
        // Build the style.
        val mediaStyle = androidx.media.app.NotificationCompat.MediaStyle()
            .setShowActionsInCompactView(0)
            .setMediaSession(mSessionToken)
        // Create/Retrieve Notification Channel for O and beyond devices (26+).
        mNotificationChannelFactory.createChannel(ServiceStartedNotificationData(mContext))
        val builder = NotificationCompat.Builder(
            mContext, ServiceStartedNotificationData.CHANNEL_ID
        )
        val smallIcon =
            if (AppUtils.hasVersionLollipop()) R.drawable.ic_notification else R.drawable.ic_notification_drawable
        builder
            .setContentIntent(makePendingIntent())
            .setStyle(mediaStyle)
            .setColor(mNotificationColor)
            .setVisibility(NotificationCompat.VISIBILITY_PUBLIC)
            .setContentTitle(mContext.getString(R.string.app_name))
            .setContentText(message)
            .setSmallIcon(smallIcon)
            .setLargeIcon(art)

        builder.addAction(getCloseAppAction())

        AppLogger.d("$CLASS_NAME show notification '$message'")
        mService.startForeground(NOTIFICATION_ID, builder.build())
    }

    /**
     * Gets the names and order of the actions to be included in the notification at the current player state.
     *
     * The playback and custom actions are combined and placed in the following order if not omitted:
     *
     * +---------------------------------------------------------------------------------+
     *
     * | prev | &lt;&lt; | play/pause | &gt;&gt; | next | custom actions | stop |
     *
     * +---------------------------------------------------------------------------------+
     *
     * The names must be of the playback actions
     * [ACTION_PAUSE], [ACTION_PLAY], [ACTION_NEXT] or [ACTION_PREV].
     * Otherwise the action name is ignored.
     */
    private fun getActions(
        enableNext: Boolean, enablePrevious: Boolean, playbackState: PlaybackStateCompat
    ): List<String> {
        val stringActions: MutableList<String> = ArrayList()
        if (enablePrevious) {
            stringActions.add(ACTION_PREV)
        }
        if (shouldShowPauseButton(playbackState)) {
            stringActions.add(ACTION_PAUSE)
        } else {
            stringActions.add(ACTION_PLAY)
        }
        if (enableNext) {
            stringActions.add(ACTION_NEXT)
        }
        stringActions.add(ACTION_CLOSE_APP)
        return stringActions
    }

    private fun getCloseAppAction(): NotificationCompat.Action {
        return NotificationCompat.Action(
            R.drawable.ic_close_app,
            mContext.getString(R.string.notif_close_app_label),
            mCloseAppIntent
        )
    }

    /**
     * Gets an array with the indices of the buttons to be shown in compact mode.
     *
     * The indices must refer to the list of actions passed as the first parameter.
     *
     * @param actionNames The names of the actions included in the notification.
     * @param playbackState State of the current playback.
     */
    private fun getActionIndicesForCompactView(
        actionNames: List<String>,
        playbackState: PlaybackStateCompat
    ): IntArray {
        val pauseActionIndex = actionNames.indexOf(ACTION_PAUSE)
        val playActionIndex = actionNames.indexOf(ACTION_PLAY)
        val skipPreviousActionIndex = if (mUseNavigationActionsInCompactView) actionNames.indexOf(ACTION_PREV) else -1
        val skipNextActionIndex = if (mUseNavigationActionsInCompactView) actionNames.indexOf(ACTION_NEXT) else -1
        val actionIndices = IntArray(4)
        var actionCounter = 0
        if (skipPreviousActionIndex != -1) {
            actionIndices[actionCounter++] = skipPreviousActionIndex
        }
        val shouldShowPauseButton = shouldShowPauseButton(playbackState)
        if (pauseActionIndex != -1 && shouldShowPauseButton) {
            actionIndices[actionCounter++] = pauseActionIndex
        } else if (playActionIndex != -1 && !shouldShowPauseButton) {
            actionIndices[actionCounter++] = playActionIndex
        }
        if (skipNextActionIndex != -1) {
            actionIndices[actionCounter++] = skipNextActionIndex
        }
        actionIndices[actionCounter++] = actionNames.indexOf(ACTION_CLOSE_APP)
        return actionIndices.copyOf(actionCounter)
    }

    private fun shouldShowPauseButton(playbackState: PlaybackStateCompat): Boolean {
        return playbackState.state != PlaybackStateCompat.STATE_PLAYING
    }

    private fun makePendingIntent(): PendingIntent? {

        // TODO: FIX ME
        val className = if (mService.isTv) "com.yuriy.openradio.tv.view.activity.TvMainActivity"
        else "com.yuriy.openradio.mobile.view.activity.MainActivity"

        val componentName = ComponentName(mContext, className)
        val intent = Intent()
        intent.component = componentName
        return PendingIntent.getActivity(mContext, 0, intent, PendingIntent.FLAG_IMMUTABLE)
    }

    private fun getPlayPauseAction(playbackState: PlaybackStateCompat): NotificationCompat.Action {
        val label: String
        val icon: Int
        val intent: PendingIntent
        if (playbackState.state == PlaybackStateCompat.STATE_PLAYING) {
            label = mContext.getString(R.string.label_pause)
            icon = R.drawable.ic_pause
            intent = mPauseIntent
        } else {
            label = mContext.getString(R.string.label_play)
            icon = R.drawable.ic_play_arrow
            intent = mPlayIntent
        }
        return NotificationCompat.Action(icon, label, intent)
    }

    companion object {
        private val CLASS_NAME = MediaNotification::class.java.simpleName
        private const val NOTIFICATION_ID = 412
        private const val ACTION_PAUSE = "com.yuriy.openradio.pause"
        private const val ACTION_PLAY = "com.yuriy.openradio.play"
        private const val ACTION_PREV = "com.yuriy.openradio.prev"
        private const val ACTION_NEXT = "com.yuriy.openradio.next"
        const val ACTION_CLOSE_APP = "com.yuriy.openradio.close_app"
    }
}
